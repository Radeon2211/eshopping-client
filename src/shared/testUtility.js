// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
