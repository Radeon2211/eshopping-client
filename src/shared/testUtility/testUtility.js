// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
