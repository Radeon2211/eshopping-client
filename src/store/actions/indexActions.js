export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions';

export { setModal, deleteMessage, changeProductsPerPage } from './uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions';
