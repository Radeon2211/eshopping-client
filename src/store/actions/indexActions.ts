export {
  registerUser,
  loginUser,
  fetchProfile,
  logoutUser,
  fetchOtherUser,
  setOtherUser,
  updateUser,
  changeEmail,
  changeDeliveryAddress,
  deleteAccount,
  addAdmin,
  removeAdmin,
  sendAccountVerificationLink,
  resetPassword,
} from './authActions/authActions';

export {
  fetchCart,
  addCartItem,
  updateCartItem,
  clearCart,
  removeCartItem,
  setTransaction,
  goToTransaction,
  buyProducts,
} from './tradeActions/tradeActions';

export { setMessage, setModal, changeProductsPerPage } from './uiActions/uiActions';

export {
  addProduct,
  editProduct,
  fetchProducts,
  fetchProductDetails,
  setProductDetails,
  deleteProduct,
} from './productActions/productActions';

export { fetchOrders, fetchOrderDetails, setOrderDetails } from './orderActions/orderActions';
