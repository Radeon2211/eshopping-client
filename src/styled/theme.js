const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
