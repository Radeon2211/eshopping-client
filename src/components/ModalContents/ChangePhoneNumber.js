import React, { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from '../../store/actions/indexActions';
import Form from '../UI/Form/Form';
import Input from '../UI/Input/Input';
import SideBySide from '../UI/SideBySide/SideBySide';
import { inputKinds } from '../../shared/constants';
import { listOfAreaCodes } from '../../shared/utility';

const validationSchema = Yup.object({
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string().max(15).trim().required(),
});

const ChangePhoneNumber = () => {
  const userProfile = useSelector((state) => state.auth.profile);

  const dispatch = useDispatch();
  const onChangePhoneNumber = useCallback((creds) => dispatch(actions.changePhoneNumber(creds)), [
    dispatch,
  ]);

  const currentPhonePrefix = userProfile.phone.split(' ')[0].split('+')[1];
  const defaultPhonePrefix = listOfAreaCodes.find(({ value }) => value === currentPhonePrefix);
  const defaultPhoneNumber = userProfile.phone.split(' ')[1];

  return (
    <Formik
      initialValues={{
        phonePrefix: defaultPhonePrefix,
        phoneNumber: defaultPhoneNumber,
      }}
      validationSchema={validationSchema}
      onSubmit={(data) => {
        onChangePhoneNumber(data);
      }}
    >
      {({ dirty, setFieldTouched, setFieldValue, isValid, touched, errors }) => (
        <Form
          btnText="change"
          headingText="Change your phone number"
          isValid={dirty && isValid}
          cancellable
        >
          <SideBySide proportion="1/1">
            <Input
              kind={inputKinds.SELECT}
              config={{
                name: 'phonePrefix',
                id: 'phonePrefix',
                value: '',
                placeholder: 'Choose your phone number prefix',
                options: listOfAreaCodes,
                setFieldValue,
                setFieldTouched,
              }}
              label="Phone number prefix"
              isValid={!errors.phonePrefix}
              isTouched={touched.phonePrefix}
            />
            <Input
              kind={inputKinds.INPUT}
              config={{
                type: 'text',
                name: 'phoneNumber',
                id: 'phoneNumber',
                placeholder: 'Your phone number',
                autoComplete: 'tel-national',
                onInput: setFieldTouched.bind(this, 'phoneNumber', true, true),
              }}
              label="Phone number"
              isValid={!errors.phoneNumber}
              isTouched={touched.phoneNumber}
            />
          </SideBySide>
        </Form>
      )}
    </Formik>
  );
};

export default ChangePhoneNumber;
