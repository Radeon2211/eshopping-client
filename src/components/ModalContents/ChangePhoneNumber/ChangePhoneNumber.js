import { useCallback } from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';
import * as actions from '../../../store/actions/indexActions';
import Form from '../../UI/Form/Form';
import Input from '../../UI/Input/Input';
import SideBySide from '../../UI/SideBySide';
import { inputKinds, userRules, listOfAreaCodes } from '../../../shared/constants';
import { getPhonePrefixAndNumber } from '../../../shared/utility/utility';

const validationSchema = Yup.object({
  phonePrefix: userRules.phonePrefix,
  phoneNumber: userRules.phoneNumber,
});

export default function ChangePhoneNumber() {
  const userProfile = useSelector((state) => state.auth.profile);

  const dispatch = useDispatch();
  const onUpdateUser = useCallback(
    (creds, message) => dispatch(actions.updateUser(creds, message)),
    [dispatch],
  );

  const { phoneNumber, phonePrefix } = getPhonePrefixAndNumber(userProfile.phone);

  return (
    <Formik
      initialValues={{
        phonePrefix,
        phoneNumber,
      }}
      validationSchema={validationSchema}
      onSubmit={(data) => {
        const correctData = { phone: `+${data.phonePrefix.value} ${data.phoneNumber}` };
        onUpdateUser(correctData, 'Phone number has been changed successfully');
      }}
    >
      {({ dirty, isValid, touched, errors, setFieldTouched, setFieldValue }) => (
        <Form
          btnText="change"
          headingText="Change your phone number"
          isValid={dirty && isValid}
          cancellable
        >
          <SideBySide $proportion="1/1">
            <Input
              kind={inputKinds.SELECT}
              config={{
                name: 'phonePrefix',
                id: 'phonePrefix',
                placeholder: 'Choose your phone number prefix',
                options: listOfAreaCodes,
                setFieldValue,
                setFieldTouched,
              }}
              label="Phone number prefix"
              isValid={!errors.phonePrefix}
              isTouched={touched.phonePrefix}
            />
            <Input
              kind={inputKinds.INPUT}
              config={{
                type: 'text',
                name: 'phoneNumber',
                id: 'phoneNumber',
                placeholder: 'Your phone number (numbers, dashes)',
                autoComplete: 'tel-national',
                autoFocus: true,
                onInput: setFieldTouched.bind(this, 'phoneNumber', true, true),
                'data-testid': 'ChangePhoneNumber-phoneNumber',
              }}
              label="Phone number"
              isValid={!errors.phoneNumber}
              isTouched={touched.phoneNumber}
            />
          </SideBySide>
        </Form>
      )}
    </Formik>
  );
}
