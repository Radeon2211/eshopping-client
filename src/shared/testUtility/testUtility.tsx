/* eslint-disable import/no-extraneous-dependencies */
import { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import moxios from 'moxios';
import rootReducer from '../../store/reducers/rootReducer';
import theme from '../../styled/theme';
import {
  CartItem,
  DeliveryAddress,
  Order,
  OrderProductItem,
  Product,
  Profile,
  ProfileStatus,
} from '../types/types';
import { ProductCondition } from '../types/enums';

export const defaultDeliveryAddress: DeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile: Profile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: ProfileStatus.ACTIVE,
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (props: Partial<Product> = {}): Product => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    quantitySold: 0,
    buyerQuantity: 0,
    description: '',
    condition: ProductCondition.NEW,
    seller: {
      username: 'username',
    },
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const createCartItem = (props = {}): CartItem => {
  const finalProps = {
    id: uuidv4(),
    sellerUsername: 'username',
    quantity: 1,
    productId: uuidv4(),
    price: 2,
    name: 'product name',
    productQuantity: 2,
    photo: false,
    ...props,
  };
  const product = createProductItem({
    _id: finalProps.productId,
    quantity: finalProps.productQuantity,
    price: finalProps.price,
    name: finalProps.name,
    photo: finalProps.photo,
    seller: {
      username: finalProps.sellerUsername,
    },
  });

  return {
    _id: finalProps.id,
    quantity: finalProps.quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  props: Partial<OrderProductItem> = {},
): OrderProductItem => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    seller: {
      username: 'username',
    },
    ...props,
  };
};

export const createOrder = (props: Partial<Order> = {}): Order => {
  return {
    _id: uuidv4(),
    overallPrice: 100,
    seller: null,
    buyer: null,
    products: [createTransactionAndOrderProdItem()],
    deliveryAddress: defaultDeliveryAddress,
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...(defaultState.product as object),
      ...initialProduct,
    },
    ui: {
      ...(defaultState.ui as object),
      ...initialUI,
    },
  };
  // @ts-expect-error TBF after redux upgrade
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

type MoxiosRequestsItems = { config: { method: string; url: string } }[];
type MockedMoxiosInstance = {
  requests: {
    __items?: MoxiosRequestsItems;
    mostRecent?: () => { config: { method: string; url: string } };
  };
};
type DefaultMoxiosInstance = Omit<typeof moxios, 'requests'> & {
  requests: typeof moxios.requests & {
    __items?: MoxiosRequestsItems;
  };
};
export const checkReqMethodAndURL = (
  moxiosInstance: MockedMoxiosInstance | DefaultMoxiosInstance,
  method: string,
  url: string,
  reqNumber = 0,
) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests?.[reqNumber - 1];
    return prevRequest && method === prevRequest.config.method && url === prevRequest.config.url;
  }
  if (!moxiosInstance.requests.mostRecent) return false;
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState: {
    auth?: object;
    product?: object;
    ui?: object;
  },
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};

export const clickAtSubmitButton = async (container: Element) => {
  await waitFor(() => {
    const submitBtn = container.querySelector('button[type="submit"]');
    if (submitBtn) {
      fireEvent.click(submitBtn);
    }
  });
};

export const testRouterPushCall = (
  pushFn: ReturnType<typeof jest.fn>,
  callIdx: number,
  expectedPathname: string,
  expectedSearch = '',
) => {
  expect(pushFn.mock.calls[callIdx][0].pathname).toEqual(expectedPathname);
  expect(pushFn.mock.calls[callIdx][0].search).toEqual(expectedSearch);
};

type RenderAppPartOptions = {
  push?: ReturnType<typeof jest.fn>;
  pathname?: string;
  search?: string;
  withoutThemeProvider?: boolean;
  withoutRouter?: boolean;
  store?: MockStoreEnhanced<object, object>;
  onlyReturnWrappedElement?: boolean;
};

export const renderAppPart = (element: ReactElement, options: RenderAppPartOptions = {}) => {
  const navigator = {
    createHref: jest.fn(),
    push: options.push || jest.fn(),
    go: jest.fn(),
    replace: jest.fn(),
  };

  const location = {
    pathname: options.pathname || '/',
    search: options.search || '',
  };

  let content = element;
  if (!options.withoutThemeProvider) {
    content = <ThemeProvider theme={theme}>{element}</ThemeProvider>;
  }
  if (options.store) {
    content = <Provider store={options.store}>{content}</Provider>;
  }
  if (!options.withoutRouter) {
    content = (
      <Router location={location} navigator={navigator}>
        {content}
      </Router>
    );
  }

  return options.onlyReturnWrappedElement ? content : render(content);
};
