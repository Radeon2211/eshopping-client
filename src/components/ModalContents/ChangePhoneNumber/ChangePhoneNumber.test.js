import React from 'react';
import { render, cleanup, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import selectEvent from 'react-select-event';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ChangePhoneNumber from './ChangePhoneNumber';
import theme from '../../../styled/theme';
import * as actions from '../../../store/actions/indexActions';
import { clickAtSubmitButton, defaultUserProfile } from '../../../shared/testUtility/testUtility';

const mockStore = configureMockStore([thunk]);

const oldPhonePrefixLabel = '+48 Poland';
const oldPhonePrefix = '+48';
const oldPhoneNumber = '123456789';

const setUp = () => {
  const store = mockStore({
    auth: {
      profile: { ...defaultUserProfile, phone: `${oldPhonePrefix} ${oldPhoneNumber}` },
    },
    ui: {
      isFormLoading: false,
      formError: '',
    },
  });
  store.dispatch = jest.fn();

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ChangePhoneNumber />
        </ThemeProvider>
      </Provider>,
    ),
    store,
  };
};

jest.mock('../../../store/actions/indexActions.js', () => ({
  ...jest.requireActual('../../../store/actions/indexActions.js'),
  updateUser: (credentials, message) => ({
    credentials,
    message,
  }),
}));

afterEach(cleanup);

describe('<ChangePhoneNumber />', () => {
  describe('Check how renders', () => {
    it('should render everything correctly', () => {
      const { asFragment } = setUp();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Check form', () => {
    const updateMessage = 'Phone number has been changed successfully';
    const defaultNewPhonePrefix = '+51';
    const defaultNewPhonePrefixLabel = '+51 Peru';
    const defaultNewPhoneNumber = '987-654-321';
    const defaultNewPhone = '+51 987-654-321';

    it('should have values from state in inputs and focus on phoneNumber input by default', () => {
      setUp();
      const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');
      expect(phoneNumberInput).toHaveFocus();
      expect(phoneNumberInput.value).toEqual(oldPhoneNumber);
      expect(screen.getByText(oldPhonePrefixLabel)).toBeInTheDocument();
    });

    describe('should call', () => {
      it('should call updateUser() with given values after input submit and button click', async () => {
        const { store, container } = setUp();

        const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');

        const dataToPass = {
          phone: defaultNewPhone,
        };

        await waitFor(() => {
          fireEvent.change(phoneNumberInput, { target: { value: defaultNewPhoneNumber } });
        });
        await waitFor(async () => {
          await selectEvent.openMenu(screen.getByText(oldPhonePrefixLabel));
          fireEvent.click(screen.getByText(defaultNewPhonePrefixLabel));
        });
        expect(phoneNumberInput.value).toEqual(defaultNewPhoneNumber);
        expect(screen.getByText(defaultNewPhonePrefixLabel)).toBeInTheDocument();

        await clickAtSubmitButton(container);
        expect(store.dispatch).toHaveBeenNthCalledWith(
          1,
          actions.updateUser(dataToPass, updateMessage),
        );

        await waitFor(() => {
          fireEvent.submit(phoneNumberInput);
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          2,
          actions.updateUser(dataToPass, updateMessage),
        );

        await waitFor(() => {
          fireEvent.submit(screen.getByText(defaultNewPhonePrefixLabel));
        });
        expect(store.dispatch).toHaveBeenNthCalledWith(
          3,
          actions.updateUser(dataToPass, updateMessage),
        );

        expect(store.dispatch).toHaveBeenCalledTimes(3);
      });

      it('should call updateUser() with changed phoneNumber only', async () => {
        const { store, container } = setUp();

        const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');

        const dataToPass = {
          phone: `${oldPhonePrefix} ${defaultNewPhoneNumber}`,
        };

        await waitFor(() => {
          fireEvent.change(phoneNumberInput, { target: { value: defaultNewPhoneNumber } });
        });
        expect(phoneNumberInput.value).toEqual(defaultNewPhoneNumber);

        await clickAtSubmitButton(container);

        expect(store.dispatch).toHaveBeenCalledWith(actions.updateUser(dataToPass, updateMessage));
      });

      it('should call updateUser() with changed phonePrefix only', async () => {
        const { store, container } = setUp();

        const dataToPass = {
          phone: `${defaultNewPhonePrefix} ${oldPhoneNumber}`,
        };

        await waitFor(async () => {
          await selectEvent.openMenu(screen.getByText(oldPhonePrefixLabel));
          fireEvent.click(screen.getByText(defaultNewPhonePrefixLabel));
        });
        expect(screen.getByText(defaultNewPhonePrefixLabel)).toBeInTheDocument();

        await clickAtSubmitButton(container);

        expect(store.dispatch).toHaveBeenCalledWith(actions.updateUser(dataToPass, updateMessage));
      });
    });

    describe('should NOT call', () => {
      it('should NOT call updateUser() if both inputs have old values', async () => {
        const { store, container } = setUp();
        await clickAtSubmitButton(container);
        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if phoneNumber is too short', async () => {
        const { store, container } = setUp();

        const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');
        const newPhoneNumber = '1234';

        await waitFor(() => {
          fireEvent.change(phoneNumberInput, { target: { value: newPhoneNumber } });
        });
        expect(phoneNumberInput.value).toEqual(newPhoneNumber);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if phoneNumber is too long', async () => {
        const { store, container } = setUp();

        const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');
        const newPhoneNumber = '123-456-789-123-2';

        await waitFor(() => {
          fireEvent.change(phoneNumberInput, { target: { value: newPhoneNumber } });
        });
        expect(phoneNumberInput.value).toEqual(newPhoneNumber);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });

      it('should NOT call updateUser() if phoneNumber also has other characters than numbers and dashes', async () => {
        const { store, container } = setUp();

        const phoneNumberInput = screen.getByTestId('ChangePhoneNumber-phone-number');
        const newPhoneNumber = '123 456 789';

        await waitFor(() => {
          fireEvent.change(phoneNumberInput, { target: { value: newPhoneNumber } });
        });
        expect(phoneNumberInput.value).toEqual(newPhoneNumber);

        await clickAtSubmitButton(container);

        expect(store.dispatch).not.toHaveBeenCalled();
      });
    });
  });
});
