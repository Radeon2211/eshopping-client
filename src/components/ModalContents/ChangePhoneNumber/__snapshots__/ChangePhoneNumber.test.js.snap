// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePhoneNumber /> check how renders should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ hYjfqi"
  >
    <h3
      class="sc-iBPRYJ kqmSLn"
    >
      Change your phone number
    </h3>
    <form
      action="#"
      class="formik-form"
    >
      <div
        class="sc-iqHYGH fkOBoS"
      >
        <div
          class="sc-fubCfw drZTsR"
        >
          <label
            class="sc-jrAGrp bkXExJ"
            for="phonePrefix"
          >
            Phone number prefix
          </label>
          <div
            class="select css-2b097c-container"
            id="phonePrefix"
          >
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-yk16xz-control"
            >
              <div
                class=" css-g1d714-ValueContainer"
              >
                <div
                  class=" css-1uccc91-singleValue"
                >
                  +48 Poland
                </div>
                <div
                  class="css-b8ldur-Input"
                >
                  <div
                    class=""
                    style="display: inline-block;"
                  >
                    <input
                      aria-autocomplete="list"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      id="react-select-2-input"
                      spellcheck="false"
                      style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1okebmr-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-tlfecz-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <input
              name="phonePrefix"
              type="hidden"
              value="48"
            />
          </div>
        </div>
        <div
          class="sc-fubCfw drZTsR"
          type="text"
        >
          <label
            class="sc-jrAGrp bkXExJ"
            for="phoneNumber"
          >
            Phone number
          </label>
          <input
            autocomplete="tel-national"
            class="sc-pFZIQ fjonmB"
            data-testid="ChangePhoneNumber-phoneNumber"
            id="phoneNumber"
            name="phoneNumber"
            placeholder="Your phone number (numbers, dashes)"
            type="text"
            value="123456789"
          />
        </div>
      </div>
      <div
        class="sc-gKsewC dLOeOj"
        direction="row"
        spacing="3"
        width="100%"
        wrap="nowrap"
      >
        <span
          class="sc-eCssSg gABWUT"
          color=""
          display="block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          <button
            class="sc-gsTCUz cpcmUx"
            color="blue"
            type="button"
          >
            cancel
          </button>
        </span>
        <button
          class="sc-gsTCUz kaTXms"
          color="blue"
          data-testid="Form-submit-btn"
          disabled=""
          type="submit"
        >
          change
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
